### ISSUE TITLE
tx hash can't be searched by history plugin,  but has contained by block？？

### ISSUE THREAD

**Issue by guaiguaihw** [EOSIO/eos#5388]
I think you should reopen the issue below, we all meet it, and it hasn't been fixed yet.
we enable filter-on=* and history plugin since the beginning, I doubt whether replay blockchain could sovle the problem
https://github.com/EOSIO/eos/issues/5334#issuecomment-415226222
@jgiszczak

**Comment by firesWu** [comment]
I think because history plugin save data through `applied_transaction` signal. so when tx is failed, trace' s action_traces is empty. so get tx by tx_id will get null. but when we get tx by block_num, it will fetch tx from block.

**Comment by taokayan** [comment]
You can use some public EOS block explorer to verify your transaction.

**Comment by guaiguaihw** [comment]
But if the exchange sync data by block, we will far behind the current block, because we can't handle one block within 0.5s. 
The only way to sync is to get action/ get transaction by account, using history plugin.
It is obvioursly that the history plugin has bugs, why don't you want to support it ?

**Comment by guaiguaihw** [comment]
The transaction was confirmed by chain, not failed, but we can't search it by get action/get transaction, even if our node is full node.
We enabled filter-on=* and history plugin since the beginning, that would not miss any index.
@firesWu

**Comment by guaiguaihw** [comment]
For example, tx_hash: 95e752dabe0bafd4b93378ec170faf23165620ca3192bb2bbb21856d43446653
It can be searched by block, but can't be searched by get transaction/ get action.
![image](https://user-images.githubusercontent.com/11972366/44565621-a4b69a00-a79b-11e8-9f22-e7ed6b1cd868.png)

**Comment by firesWu** [comment]
You can see tx json data from block. traces is null. If tx executed successfully, traces shoud have data. so that we cannot get tx by tx_id, because traces is null and history_plugin parse tx through traces

```
{
    "id": "95e752dabe0bafd4b93378ec170faf23165620ca3192bb2bbb21856d43446653",
    "trx": {
        "receipt": {
            "status": "executed",
            "cpu_usage_us": 1333,
            "net_usage_words": 17,
            "trx": [
                1,
                {
                    "signatures": [
                        "SIG_K1_KiUbAgztj1NySrpCo6MhzuPhXNp4gCsEJuDmwm4yA5P9BrZD5qnCP3xPdRmuuMniCvgJCLNaSzoikDemwD6SyLNWxszy5A"
                    ],
                    "compression": "none",
                    "packed_context_free_data": "",
                    "packed_trx": "fd4e625ba4ea99c5614e000000000100a6823403ea3055000000572d3ccdcd0170b4a29d6aefa6de00000000a8ed32322770b4a29d6aefa6de302995aa659a0fa4104a0f000000000004454f53000000000636393235313000"
                }
            ]
        },
        "trx": {
            "expiration": "2018-08-02T00:23:25",
            "ref_block_num": 60068,
            "ref_block_prefix": 1315030425,
            "max_net_usage_words": 0,
            "max_cpu_usage_ms": 0,
            "delay_sec": 0,
            "context_free_actions": [],
            "actions": [
                {
                    "account": "eosio.token",
                    "name": "transfer",
                    "authorization": [
                        {
                            "actor": "vuniyuoxoeub",
                            "permission": "active"
                        }
                    ],
                    "data": {
                        "from": "vuniyuoxoeub",
                        "to": "okbtothemoon",
                        "quantity": "100.2000 EOS",
                        "memo": "692510"
                    },
                    "hex_data": "70b4a29d6aefa6de302995aa659a0fa4104a0f000000000004454f530000000006363932353130"
                }
            ],
            "transaction_extensions": [],
            "signatures": [
                "SIG_K1_KiUbAgztj1NySrpCo6MhzuPhXNp4gCsEJuDmwm4yA5P9BrZD5qnCP3xPdRmuuMniCvgJCLNaSzoikDemwD6SyLNWxszy5A"
            ],
            "context_free_data": []
        }
    },
    "block_time": "2018-08-01T23:52:57.500",
    "block_num": 8973302,
    "last_irreversible_block": 12740862,
    "traces": []
}
```

**Comment by guaiguaihw** [comment]
![image](https://user-images.githubusercontent.com/11972366/44565899-370b6d80-a79d-11e8-968b-8c1c6ed9fdae.png)
tx: 95e752dabe0bafd4b93378ec170faf23165620ca3192bb2bbb21856d43446653

**Comment by guaiguaihw** [comment]
So you think the transaction did not executed successfully？But get balance shows the amount has been successfully transfer to our account. Furthermore, eosio has no type to define if the transaction has been executed successfully or not, you can't judge it by traces. @firesWu

**Comment by firesWu** [comment]
transaction_trace is different from transaction_metadata.  trx is telling block that you need to execute these tx. traces  is the execution result of tx. so when tx failed in execution. we can see tx in trx, but we cannot see tx in traces and history_plugin store data through traces.  I show you a example. tx_id is `6d61113404e27ba26f13f67d447aed435f83b9e705d8f0ddabf953a078089701` , block_num is `511019`
and traces result is
```
{
  "id": "6d61113404e27ba26f13f67d447aed435f83b9e705d8f0ddabf953a078089701",
  "receipt": {
    "status": "hard_fail",
    "cpu_usage_us": 0,
    "net_usage_words": 0
  },
  "elapsed": 521,
  "net_usage": 0,
  "scheduled": true,
  "action_traces": [
    {
      "receipt": {
        "receiver": "",
        "act_digest": "0000000000000000000000000000000000000000000000000000000000000000",
        "global_sequence": 0,
        "recv_sequence": 0,
        "auth_sequence": [
          
        ],
        "code_sequence": 0,
        "abi_sequence": 0
      },
      "act": {
        "account": "",
        "name": "",
        "authorization": [
          
        ],
        "data": ""
      },
      "elapsed": 0,
      "cpu_usage": 0,
      "console": "",
      "total_cpu_usage": 0,
      "trx_id": "0000000000000000000000000000000000000000000000000000000000000000",
      "inline_traces": [
        
      ]
    }
  ],
  "failed_dtrx_trace": null,
  "except": {
    "code": 3050003,
    "name": "eosio_assert_message_exception",
    "message": "eosio_assert_message assertion failure",
    "stack": [
      {
        "context": {
          "level": "error",
          "file": "wasm_interface.cpp",
          "line": 930,
          "method": "eosio_assert",
          "hostname": "",
          "thread_name": "thread-0",
          "timestamp": "2018-08-14T03:45:06.355"
        },
        "format": "assertion failure with message: ${s}",
        "data": {
          "s": "refund is not available yet"
        }
      },
      {
        "context": {
          "level": "warn",
          "file": "apply_context.cpp",
          "line": 61,
          "method": "exec_one",
          "hostname": "",
          "thread_name": "thread-0",
          "timestamp": "2018-08-14T03:45:06.356"
        },
        "format": "pending console output: ${console}",
        "data": {
          "console": ""
        }
      }
    ]
  }
}
```

**Comment by guaiguaihw** [comment]
But `cleos get currency balance` shows that the transaction has been executed successfully, we get the tokens, and the from account reduce tokens. It make us so confused. Which result should 
I believe？I think this must be a bug that should be supported immediately
@firesWu

**Comment by guaiguaihw** [comment]
Besides, I doubt the way you judge whether the transaction has been executed successfully or not. We have already send some transactions which were not executed successfully, the tx_id wouldn't be found in blocks finally. @firesWu

**Comment by winlin** [comment]
should we not use the `status` in the transaction to judge whether an transaction success or failure ? @firesWu

**Comment by heifner** [comment]
@guaiguaihw The transaction is available here:
https://eospark.com/MainNet/tx/95e752dabe0bafd4b93378ec170faf23165620ca3192bb2bbb21856d43446653

I believe a replay with the current version of eos.io would show that transaction.

@firesWu I'm planning on adding `action_trace` for `hard_fail` so that it can be tracked via `action_trace` without having to look at the `transaction_trace`. This is useful for the `mongo_db_plugin` so that `transaction_traces` do not have to be stored, only `action_traces` because I think a common use case will be to store `action_traces` and `block_states` only.

In mongo_db_plugin for `irreversible` certainty you can check if `trx_id` of `action_trace` is in `block_state` with `irreversible = true`.

**Comment by heifner** [comment]
@guaiguaihw The transaction is in my local mongo database:

```
> db.action_traces.find({"trx_id":"95e752dabe0bafd4b93378ec170faf23165620ca3192bb2bbb21856d43446653"}).pretty()
{
	"_id" : ObjectId("5b7e9867f9a2f323e822f3a4"),
	"receipt" : {
		"receiver" : "eosio.token",
		"act_digest" : "8204055e1bdb36a685190cb8be16d106b0bbc5786175569d21a342bfa814af82",
		"global_sequence" : 144659376,
		"recv_sequence" : 2607016,
		"auth_sequence" : [
			[
				"vuniyuoxoeub",
				46056
			]
		],
		"code_sequence" : 2,
		"abi_sequence" : 2
	},
	"act" : {
		"account" : "eosio.token",
		"name" : "transfer",
		"authorization" : [
			{
				"actor" : "vuniyuoxoeub",
				"permission" : "active"
			}
		],
		"data" : {
			"from" : "vuniyuoxoeub",
			"to" : "okbtothemoon",
			"quantity" : "100.2000 EOS",
			"memo" : "692510"
		},
		"hex_data" : "70b4a29d6aefa6de302995aa659a0fa4104a0f000000000004454f530000000006363932353130"
	},
	"elapsed" : 267,
	"cpu_usage" : 0,
	"console" : "",
	"total_cpu_usage" : 0,
	"trx_id" : "95e752dabe0bafd4b93378ec170faf23165620ca3192bb2bbb21856d43446653",
	"block_num" : 8973302,
	"block_time" : "2018-08-01T23:52:57.500",
	"producer_block_id" : "0088ebf6687bfb7f1230d8d36d632f1f0ef6fc27a66af160c6c94449e1220b1c",
	"account_ram_delta" : [ ],
	"trx_status" : "executed",
	"createdAt" : ISODate("2018-08-23T11:20:07.972Z")
}
{
	"_id" : ObjectId("5b7e9867f9a2f323e822f3a5"),
	"receipt" : {
		"receiver" : "vuniyuoxoeub",
		"act_digest" : "8204055e1bdb36a685190cb8be16d106b0bbc5786175569d21a342bfa814af82",
		"global_sequence" : 144659377,
		"recv_sequence" : 23123,
		"auth_sequence" : [
			[
				"vuniyuoxoeub",
				46057
			]
		],
		"code_sequence" : 2,
		"abi_sequence" : 2
	},
	"act" : {
		"account" : "eosio.token",
		"name" : "transfer",
		"authorization" : [
			{
				"actor" : "vuniyuoxoeub",
				"permission" : "active"
			}
		],
		"data" : {
			"from" : "vuniyuoxoeub",
			"to" : "okbtothemoon",
			"quantity" : "100.2000 EOS",
			"memo" : "692510"
		},
		"hex_data" : "70b4a29d6aefa6de302995aa659a0fa4104a0f000000000004454f530000000006363932353130"
	},
	"elapsed" : 4,
	"cpu_usage" : 0,
	"console" : "",
	"total_cpu_usage" : 0,
	"trx_id" : "95e752dabe0bafd4b93378ec170faf23165620ca3192bb2bbb21856d43446653",
	"block_num" : 8973302,
	"block_time" : "2018-08-01T23:52:57.500",
	"producer_block_id" : "0088ebf6687bfb7f1230d8d36d632f1f0ef6fc27a66af160c6c94449e1220b1c",
	"account_ram_delta" : [ ],
	"trx_status" : "executed",
	"createdAt" : ISODate("2018-08-23T11:20:07.972Z")
}
{
	"_id" : ObjectId("5b7e9867f9a2f323e822f3a6"),
	"receipt" : {
		"receiver" : "okbtothemoon",
		"act_digest" : "8204055e1bdb36a685190cb8be16d106b0bbc5786175569d21a342bfa814af82",
		"global_sequence" : 144659378,
		"recv_sequence" : 9695,
		"auth_sequence" : [
			[
				"vuniyuoxoeub",
				46058
			]
		],
		"code_sequence" : 2,
		"abi_sequence" : 2
	},
	"act" : {
		"account" : "eosio.token",
		"name" : "transfer",
		"authorization" : [
			{
				"actor" : "vuniyuoxoeub",
				"permission" : "active"
			}
		],
		"data" : {
			"from" : "vuniyuoxoeub",
			"to" : "okbtothemoon",
			"quantity" : "100.2000 EOS",
			"memo" : "692510"
		},
		"hex_data" : "70b4a29d6aefa6de302995aa659a0fa4104a0f000000000004454f530000000006363932353130"
	},
	"elapsed" : 5,
	"cpu_usage" : 0,
	"console" : "",
	"total_cpu_usage" : 0,
	"trx_id" : "95e752dabe0bafd4b93378ec170faf23165620ca3192bb2bbb21856d43446653",
	"block_num" : 8973302,
	"block_time" : "2018-08-01T23:52:57.500",
	"producer_block_id" : "0088ebf6687bfb7f1230d8d36d632f1f0ef6fc27a66af160c6c94449e1220b1c",
	"account_ram_delta" : [ ],
	"trx_status" : "executed",
	"createdAt" : ISODate("2018-08-23T11:20:07.972Z")
}
```

**Comment by guaiguaihw** [comment]
Replay cost almost two day, if we enabled filter-on=*.

**Comment by jgiszczak** [comment]
Defects in history_plugin were corrected with release v1.4.3.  See #6220 for details.  A replay with version 1.4.3 or later should resolve the discrepancies reported in this issue.
